File: ebay-product-marketing-analysis-search/ebay-product-search/src/main/java/org/example/message/Product.java
package org.example.message;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Pattern;
import lombok.Data;


@Data
@Entity
@Table(name = "product")
public class Product {
    @NotBlank(message = "uniqId is required")
    @NotNull
    @Id
    @Column(name = "uniq_id", nullable = false, length = 32)
    private String uniqId;           // 全局唯一ID
    @NotBlank(message = "crawlTimestamp is required")
    @Pattern(regexp = "\\d{4}-\\d{2}-\\d{2} \\d{2}:\\d{2}:\\d{2}",
            message = "Invalid crawlTimestamp format")
    @Column(name = "crawl_timestamp", nullable = false)
    private String crawlTimestamp;   // 爬取时间，格式为：yyyy-MM-dd HH:mm:ss
    @NotBlank(message = "pageUrl is required")
    @Column(name = "page_url", nullable = false)
    private String pageUrl;          // 产品页面URL
    @NotBlank(message = "title is required")
    @Column(name = "title", nullable = false)
    private String title;            // 产品标题
    @NotBlank(message = "modelNum is required")
    @Column(name = "model_num", nullable = false)
    private String modelNum;         // 产品型号
    @NotBlank(message = "manufacturer is required")
    @Column(name = "manufacturer", nullable = false)
    private String manufacturer;     // 厂商名称
    @NotBlank(message = "modelName is required")
    @Column(name = "model_name", nullable = false)
    private String modelName;        // 产品名称
    @NotBlank(message = "price is required")
    @Column(name = "price", nullable = false)
    private String price;            // 价格
    @NotNull
    @Column(name = "stock", nullable = false)
    private Integer stock;           // 是否有库存（0: 无，1: 有）
    @NotBlank(message = "carrier is required")
    @Column(name = "carrier", nullable = false)
    private String carrier;          // 运营商（如AT&T）
    @NotBlank(message = "colorCategory is required")
    @Column(name = "color_category", nullable = false)
    private String colorCategory;    // 颜色分类
    @NotBlank(message = "internalMemory is required")
    @Column(name = "internal_memory", nullable = false)
    private String internalMemory;   // 内存配置（如 64 GB）
    @NotBlank(message = "screenSize is required")
    @Column(name = "screen_size", nullable = false)
    private String screenSize;       // 屏幕尺寸（如 6.7 in）
    @NotNull
    @Column(name = "discontinued", nullable = false)
    private Integer discontinued;    // 是否停售（0: 未停售，1: 停售）
    @NotNull
    @Column(name = "broken_link", nullable = false)
    private Integer brokenLink;      // 链接是否失效（0: 未失效，1: 失效）
    @NotNull
    @Column(name = "seller_rating", nullable = false)
    private Double sellerRating;     // 卖家评分
    @NotNull
    @Column(name = "seller_num_of_reviews", nullable = false)
    private Integer sellerNumOfReviews; // 卖家评价数量
    @NotNull
    @Column(name = "average_star", nullable = false)
    private Double averageStar;      // 平均评分
}

File: ebay-product-marketing-analysis-search/product-message-sending/src/product.rs
use serde::{Deserialize, Serialize, Serializer};


#[derive(Debug, Deserialize)]
pub struct Product {
    pub uniq_id: String,
    pub crawl_timestamp: String,
    pub page_url: String,
    pub title: String,
    pub model_num: String,
    pub manufacturer: String,
    pub model_name: String,
    pub price: String,
    pub stock: i8,
    pub carrier: String,
    pub color_category: String,
    pub internal_memory: String,
    pub screen_size: String,
    pub discontinued: i8,
    pub broken_link: i8,
    pub seller_rating: f64,
    pub seller_num_of_reviews: i32,
    pub average_star: f64,
}

fn serialize_as_camel_case<S>(product: &Product, serializer: S) -> Result<S::Ok, S::Error>
where
    S: Serializer,
{
    use serde::ser::SerializeStruct;
    let mut state = serializer.serialize_struct("Product", 18)?;
    state.serialize_field("uniqId", &product.uniq_id)?;
    state.serialize_field("crawlTimestamp", &product.crawl_timestamp)?;
    state.serialize_field("pageUrl", &product.page_url)?;
    state.serialize_field("title", &product.title)?;
    state.serialize_field("modelNum", &product.model_num)?;
    state.serialize_field("manufacturer", &product.manufacturer)?;
    state.serialize_field("modelName", &product.model_name)?;
    state.serialize_field("price", &product.price)?;
    state.serialize_field("stock", &product.stock)?;
    state.serialize_field("carrier", &product.carrier)?;
    state.serialize_field("colorCategory", &product.color_category)?;
    state.serialize_field("internalMemory", &product.internal_memory)?;
    state.serialize_field("screenSize", &product.screen_size)?;
    state.serialize_field("discontinued", &product.discontinued)?;
    state.serialize_field("brokenLink", &product.broken_link)?;
    state.serialize_field("sellerRating", &product.seller_rating)?;
    state.serialize_field("sellerNumOfReviews", &product.seller_num_of_reviews)?;
    state.serialize_field("averageStar", &product.average_star)?;
    state.end()
}

impl Serialize for Product {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize_as_camel_case(self, serializer)
    }
}

impl Product {
    pub fn remove_timezone_suffix(&mut self) {
        match self.crawl_timestamp.rfind(" ") {
            Some(index) => {
                self.crawl_timestamp = self.crawl_timestamp[..index].to_string();
            }
            None => panic!("Failed to remove timezone suffix from crawl_timestamp: {}", self.crawl_timestamp)
            
        }
    }
}
